File: src/ravel/knowledge.py
"""Fail-closed knowledge promotion. Rust is the canonical implementation."""

from __future__ import annotations

from dataclasses import asdict, dataclass
from typing import Mapping, Sequence


STAGES = {
    "observation": ("episode",),
    "episode": ("open_hypothesis", "retired"),
    "open_hypothesis": ("intervention", "counterexample", "retired"),
    "intervention": ("attribution", "counterexample", "retired"),
    "attribution": ("provisional_principle", "counterexample", "retired"),
    "provisional_principle": ("transfer_tested_principle", "counterexample", "retired"),
    "transfer_tested_principle": ("restricted_strategy", "counterexample", "retired"),
    "restricted_strategy": ("supported_strategy", "counterexample", "retired"),
    "supported_strategy": ("counterexample", "retired"),
    "counterexample": (),
    "retired": (),
}


class KnowledgeError(ValueError):
    """Raised when a knowledge transition would skip evidence or rewrite status."""


@dataclass(frozen=True, slots=True)
class AttributionRecord:
    attribution_id: str
    source_intervention_ids: tuple[str, ...]
    evaluator_identity: str
    evidence_ids: tuple[str, ...]
    disposition: str
    scope: Mapping[str, str]


@dataclass(frozen=True, slots=True)
class TransferTestRecord:
    test_id: str
    principle_id: str
    context_identity: str
    evidence_ids: tuple[str, ...]
    outcome: str


@dataclass(frozen=True, slots=True)
class KnowledgeRecord:
    record_id: str
    stage: str
    statement: str
    scope: Mapping[str, str]
    parent_ids: tuple[str, ...] = ()
    evidence_ids: tuple[str, ...] = ()
    evaluation_status: str | None = None
    transfer_status: str = "untested"
    attribution_id: str | None = None
    transfer_test_ids: tuple[str, ...] = ()
    challenged_ids: tuple[str, ...] = ()
    producer_id: str = "ravel-knowledge"
    created_at: str = ""

    def to_dict(self) -> dict[str, object]:
        payload = asdict(self)
        payload["scope"] = dict(self.scope)
        payload["parent_ids"] = list(self.parent_ids)
        payload["evidence_ids"] = list(self.evidence_ids)
        payload["transfer_test_ids"] = list(self.transfer_test_ids)
        payload["challenged_ids"] = list(self.challenged_ids)
        return payload


def promote(
    current: KnowledgeRecord,
    *,
    next_stage: str,
    next_id: str,
    statement: str,
    evidence_ids: tuple[str, ...] = (),
    evaluation_status: str | None = None,
    attribution: AttributionRecord | None = None,
    transfer_tests: Sequence[TransferTestRecord] = (),
    created_at: str,
) -> KnowledgeRecord:
    """Create the next knowledge record or fail closed."""

    if next_stage not in STAGES.get(current.stage, ()):
        raise KnowledgeError(
            f"invalid knowledge transition: {current.stage}->{next_stage}"
        )
    if current.evaluation_status in {"UNKNOWN", "FAIL"} and evaluation_status == "PASS":
        raise KnowledgeError("knowledge promotion cannot convert FAIL or UNKNOWN into PASS")
    if next_id == current.record_id:
        raise KnowledgeError("promotion must not overwrite its parent")
    if next_stage == "provisional_principle":
        if attribution is None or attribution.disposition != "supported":
            raise KnowledgeError("a principle requires a valid attribution record")
        if dict(attribution.scope) != dict(current.scope):
            raise KnowledgeError("attribution scope must match the parent knowledge scope")
    if next_stage == "transfer_tested_principle" and not transfer_tests:
        raise KnowledgeError("a transfer-tested principle must identify actual transfer tests")
    if next_stage == "supported_strategy":
        contexts = {
            item.context_identity for item in transfer_tests if item.outcome == "supported"
        }
        if len(contexts) < 2:
            raise KnowledgeError(
                "a supported strategy must not be created from one local context"
            )
    if next_stage == "counterexample" and not evidence_ids:
        raise KnowledgeError(
            "a counterexample must remain linked to the knowledge it challenges"
        )
    return KnowledgeRecord(
        record_id=next_id,
        stage=next_stage,
        statement=statement,
        scope=dict(current.scope),
        parent_ids=(current.record_id,),
        evidence_ids=evidence_ids,
        evaluation_status=evaluation_status,
        transfer_status="supported" if next_stage == "supported_strategy" else "untested",
        attribution_id=None if attribution is None else attribution.attribution_id,
        transfer_test_ids=tuple(item.test_id for item in transfer_tests),
        challenged_ids=(current.record_id,) if next_stage == "counterexample" else (),
        producer_id=current.producer_id,
        created_at=created_at,
    )

File: crates/ravel-memory/src/knowledge.rs
//! Fail-closed knowledge promotion. Source records stay immutable.

use ravel_contracts::canonical_json;
use ravel_contracts::identity::hex_sha256;
use ravel_contracts::schema::KNOWLEDGE_RECORD_SCHEMA;
use ravel_contracts::status::EvidenceStatus;
use serde::{Deserialize, Serialize};
use serde_json::{Value, json};
use std::collections::BTreeMap;
use thiserror::Error;

#[derive(Debug, Error)]
pub enum KnowledgeError {
    #[error("{0}")]
    Invalid(String),
    #[error(transparent)]
    Canonical(#[from] ravel_contracts::CanonicalError),
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum KnowledgeStage {
    Observation,
    Episode,
    OpenHypothesis,
    Intervention,
    Attribution,
    ProvisionalPrinciple,
    TransferTestedPrinciple,
    RestrictedStrategy,
    SupportedStrategy,
    Counterexample,
    Retired,
}

impl KnowledgeStage {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Observation => "observation",
            Self::Episode => "episode",
            Self::OpenHypothesis => "open_hypothesis",
            Self::Intervention => "intervention",
            Self::Attribution => "attribution",
            Self::ProvisionalPrinciple => "provisional_principle",
            Self::TransferTestedPrinciple => "transfer_tested_principle",
            Self::RestrictedStrategy => "restricted_strategy",
            Self::SupportedStrategy => "supported_strategy",
            Self::Counterexample => "counterexample",
            Self::Retired => "retired",
        }
    }

    pub fn parse(value: &str) -> Result<Self, KnowledgeError> {
        match value {
            "observation" => Ok(Self::Observation),
            "episode" => Ok(Self::Episode),
            "open_hypothesis" => Ok(Self::OpenHypothesis),
            "intervention" => Ok(Self::Intervention),
            "attribution" => Ok(Self::Attribution),
            "provisional_principle" => Ok(Self::ProvisionalPrinciple),
            "transfer_tested_principle" => Ok(Self::TransferTestedPrinciple),
            "restricted_strategy" => Ok(Self::RestrictedStrategy),
            "supported_strategy" => Ok(Self::SupportedStrategy),
            "counterexample" => Ok(Self::Counterexample),
            "retired" => Ok(Self::Retired),
            other => Err(KnowledgeError::Invalid(format!(
                "unknown knowledge stage: {other}"
            ))),
        }
    }

    fn allowed_next(self) -> &'static [Self] {
        match self {
            Self::Observation => &[Self::Episode],
            Self::Episode => &[Self::OpenHypothesis, Self::Retired],
            Self::OpenHypothesis => &[Self::Intervention, Self::Counterexample, Self::Retired],
            Self::Intervention => &[Self::Attribution, Self::Counterexample, Self::Retired],
            Self::Attribution => &[
                Self::ProvisionalPrinciple,
                Self::Counterexample,
                Self::Retired,
            ],
            Self::ProvisionalPrinciple => &[
                Self::TransferTestedPrinciple,
                Self::Counterexample,
                Self::Retired,
            ],
            Self::TransferTestedPrinciple => &[
                Self::RestrictedStrategy,
                Self::Counterexample,
                Self::Retired,
            ],
            Self::RestrictedStrategy => {
                &[Self::SupportedStrategy, Self::Counterexample, Self::Retired]
            }
            Self::SupportedStrategy => &[Self::Counterexample, Self::Retired],
            Self::Counterexample | Self::Retired => &[],
        }
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct KnowledgeRecord {
    pub record_id: String,
    pub stage: KnowledgeStage,
    pub statement: String,
    pub scope: BTreeMap<String, String>,
    pub parent_ids: Vec<String>,
    pub evidence_ids: Vec<String>,
    pub evaluation_status: Option<EvidenceStatus>,
    pub transfer_status: String,
    pub attribution: Option<String>,
    pub producer_id: String,
    pub created_at: String,
}

impl KnowledgeRecord {
    pub fn validate(&self) -> Result<(), KnowledgeError> {
        if self.record_id.trim().is_empty() || self.statement.trim().is_empty() {
            return Err(KnowledgeError::Invalid(
                "knowledge record identity and statement are required".into(),
            ));
        }
        if self.scope.is_empty() {
            return Err(KnowledgeError::Invalid(
                "knowledge record scope is required".into(),
            ));
        }
        Ok(())
    }

    pub fn to_value(&self) -> Value {
        json!({
            "schema": KNOWLEDGE_RECORD_SCHEMA,
            "record_id": self.record_id,
            "stage": self.stage.as_str(),
            "statement": self.statement,
            "scope": self.scope,
            "parent_ids": self.parent_ids,
            "evidence_ids": self.evidence_ids,
            "evaluation_status": self.evaluation_status.map(|item| item.as_str()),
            "transfer_status": self.transfer_status,
            "attribution": self.attribution,
            "producer_id": self.producer_id,
            "created_at": self.created_at,
        })
    }

    pub fn digest(&self) -> Result<String, KnowledgeError> {
        Ok(format!(
            "sha256:{}",
            hex_sha256(canonical_json(&self.to_value())?.as_bytes())
        ))
    }
}

#[allow(clippy::too_many_arguments)]
pub fn promote(
    current: &KnowledgeRecord,
    next_stage: KnowledgeStage,
    next_id: &str,
    statement: &str,
    evidence_ids: Vec<String>,
    evaluation_status: Option<EvidenceStatus>,
    transfer_status: &str,
    attribution: Option<String>,
    created_at: &str,
) -> Result<KnowledgeRecord, KnowledgeError> {
    current.validate()?;
    if !current.stage.allowed_next().contains(&next_stage) {
        return Err(KnowledgeError::Invalid(format!(
            "invalid knowledge transition: {}->{}",
            current.stage.as_str(),
            next_stage.as_str()
        )));
    }
    if current.stage == KnowledgeStage::Episode && next_stage == KnowledgeStage::SupportedStrategy {
        return Err(KnowledgeError::Invalid(
            "an episode cannot directly become a global strategy".into(),
        ));
    }
    if next_stage == KnowledgeStage::ProvisionalPrinciple
        && attribution.as_deref() != Some("supported")
    {
        return Err(KnowledgeError::Invalid(
            "a successful intervention cannot become a principle without supported attribution"
                .into(),
        ));
    }
    if next_stage == KnowledgeStage::TransferTestedPrinciple && evidence_ids.is_empty() {
        return Err(KnowledgeError::Invalid(
            "an untested principle cannot authorize transfer".into(),
        ));
    }
    if next_stage == KnowledgeStage::SupportedStrategy && transfer_status != "supported" {
        return Err(KnowledgeError::Invalid(
            "an untested principle cannot authorize broad reuse".into(),
        ));
    }
    if current.evaluation_status == Some(EvidenceStatus::Unknown)
        && evaluation_status == Some(EvidenceStatus::Pass)
    {
        return Err(KnowledgeError::Invalid(
            "knowledge promotion cannot convert UNKNOWN into PASS".into(),
        ));
    }
    if next_stage == KnowledgeStage::Counterexample && evidence_ids.is_empty() {
        return Err(KnowledgeError::Invalid(
            "a failed transfer test must remain linked to the principle".into(),
        ));
    }
    let promoted = KnowledgeRecord {
        record_id: next_id.to_string(),
        stage: next_stage,
        statement: statement.to_string(),
        scope: current.scope.clone(),
        parent_ids: vec![current.record_id.clone()],
        evidence_ids,
        evaluation_status,
        transfer_status: transfer_status.to_string(),
        attribution,
        producer_id: current.producer_id.clone(),
        created_at: created_at.to_string(),
    };
    promoted.validate()?;
    Ok(promoted)
}

#[cfg(test)]
#[allow(clippy::expect_used, clippy::unwrap_used)]
mod tests {
    use super::*;

    fn observation() -> KnowledgeRecord {
        KnowledgeRecord {
            record_id: "knowledge:obs".into(),
            stage: KnowledgeStage::Observation,
            statement: "A bounded Forge check returned UNKNOWN.".into(),
            scope: BTreeMap::from([("partition".into(), "development".into())]),
            parent_ids: Vec::new(),
            evidence_ids: vec!["obs:1".into()],
            evaluation_status: Some(EvidenceStatus::Unknown),
            transfer_status: "untested".into(),
            attribution: None,
            producer_id: "ravel-knowledge".into(),
            created_at: "2026-08-14T00:00:00Z".into(),
        }
    }

    #[test]
    fn skips_and_unknown_to_pass_fail_closed() {
        let current = observation();
        assert!(
            promote(
                &current,
                KnowledgeStage::SupportedStrategy,
                "knowledge:bad",
                "must not skip",
                vec!["e".into()],
                Some(EvidenceStatus::Unknown),
                "supported",
                None,
                "t1",
            )
            .is_err()
        );
        let episode = promote(
            &current,
            KnowledgeStage::Episode,
            "knowledge:ep",
            &current.statement,
            current.evidence_ids.clone(),
            Some(EvidenceStatus::Unknown),
            "untested",
            None,
            "t1",
        )
        .expect("episode");
        assert!(
            promote(
                &episode,
                KnowledgeStage::Episode,
                "knowledge:pass",
                &episode.statement,
                episode.evidence_ids.clone(),
                Some(EvidenceStatus::Pass),
                "untested",
                None,
                "t2",
            )
            .is_err()
        );
    }
}
